python/collision_checker.py

```diff
@@ -27,22 +30,51 @@ def GPS_callback(self, data):
     def AIS_callback(self, data):
         self.ships = data.ships
 
-    def check_for_collisions(self):
-        for ship in self.ships:
-            dist = distance((ship.lat, ship.lon), (self.lat, self.lon)).km
-            if dist < self.collision_radius:
-                rospy.logfatal("Boat has collided with obstacle. Collision radius {}km. Actual distance to boat: {}km"
-                               .format(self.collision_radius, dist))
+    def check_for_collisions(self, colliding_ship):
+        check_ship = colliding_ship
 
-            elif dist < self.warn_radius:
-                rospy.logwarn("Close to collision. Within {}km of boat. Actual distance to boat: {}km"
-                              .format(self.warn_radius, dist))
+        if check_ship is None:
+            for ship in self.ships:
+                dist = distance((ship.lat, ship.lon), (self.lat, self.lon)).km
+
+                if dist < self.collision_radius:
+                    check_ship = ship
+                    self.times_collided = self.times_collided + 1
+                    rospy.logfatal("Boat has collided with obstacle. Collision radius {}km.\n"
+                                   "\tActual distance to boat: {}km. Collision number: {}"
+                                   .format(self.collision_radius, dist, self.times_collided))
+                    utilities.takeScreenshot()
+
+                elif dist < self.warn_radius:
+                    check_ship = ship
+                    self.times_warned = self.times_warned + 1
+                    rospy.logwarn("Close to collision. Within {}km of boat.\n\tActual distance to boat: {}km. "
+                                  "Warning number: {}" .format(self.warn_radius, dist, self.times_warned))
+                    utilities.takeScreenshot()
+        else:
+            dist = distance((check_ship.lat, check_ship.lon), (self.lat, self.lon)).km
+            if not (dist < self.collision_radius or dist < self.warn_radius):
+                check_ship = None
+                rospy.logwarn("Other ships are not within the collision or warning radius anymore.")
+
+        return check_ship
+
+    def get_times_collided(self):
+        return self.times_collided
+
+    def get_times_warned(self):
+        return self.times_warned
 
 
 if __name__ == '__main__':
     collision_checker = CollisionChecker(COLLISION_RADIUS_KM, WARN_RADIUS_KM)
+
     rate = rospy.Rate(1 / CHECK_PERIOD)
 
+    shipColliding = None
```

Comment:
I think the logic could be improved by removing the use of shipColliding and replacing it with a class variable liist/set in CollisionChecker. Eg.`self.currentlyCollidingBoats` and `self.currentlyWarningBoats`.

Then in your for loop looking at boats, you can check if the boat is in `self.currentlyCollidingBoats` or `self.currentlyWarningBoats`.

If it is in that list, then it is the same boat. No new collision. If it is a new boat, then add it the list and increment the counter. You can also check the number of boats in `self.currentlyCollidingBoats` or `self.currentlyWarningBoats` that were seen in the for loop.

Let me know if that makes sense or you feel something else would be better. This is just my suggestion on how to ensure this code has behavior we want.